File: packages/fast-sync-core/src/core/implementation/service/conflicts_handler.ts
import { ISyncableObject } from "../../abstraction/metadata/ISyncable_object";
import { IConflictsHandler } from "../../abstraction/service/IConflicts_handler";
import { ConflictsResolutionStrategyEnum } from "../../abstraction/service/conflicts_resolution_strategie";
import { isNullOrUndefined } from "../utils/utils";

export class ConflictsHandler implements IConflictsHandler{

    constructor(private resolutionStrategy: ConflictsResolutionStrategyEnum = ConflictsResolutionStrategyEnum.LastWriterWins, 
        private conflictsResolutionFunction?: (oblObject: ISyncableObject, newObject: ISyncableObject) => Promise<ISyncableObject>){

        if (ConflictsResolutionStrategyEnum.PredefinedRules == this.resolutionStrategy && isNullOrUndefined(conflictsResolutionFunction)) {
            throw new Error("conflictsResolutionFunction is required when you put ConflictsResolutionStrategyEnum.PredefinedRules");
        }
    }

    public getConflictsResolutionStrategy(){
        return this.resolutionStrategy;
    }

    public async resolveConflicts(oblObject: ISyncableObject, newObject: ISyncableObject) :Promise<ISyncableObject>{
        if(ConflictsResolutionStrategyEnum.LastWriterWins == this.resolutionStrategy){
            return newObject;
        }

        if(ConflictsResolutionStrategyEnum.TimestampOrdering == this.resolutionStrategy){
            return oblObject.metadata.timestamp > newObject.metadata.timestamp ? oblObject : newObject;
        }

        if(ConflictsResolutionStrategyEnum.PredefinedRules == this.resolutionStrategy){
           let resolvedobject = await this.conflictsResolutionFunction(oblObject, newObject);
           resolvedobject.metadata = newObject.metadata;
           return resolvedobject
        }
    }
}

File: fast_sync_dart_client/fast_sync_core/lib/src/implementation/service/conflicts_handler.dart
import 'package:fast_sync_client/src/absraction/metadata/isyncalbe_object.dart';
import 'package:fast_sync_client/src/absraction/service/conflicts_resolution_strategie.dart';
import 'package:fast_sync_client/src/absraction/service/iconflicts_handler.dart';

class ConflictsHandler implements IConflictsHandler {
  ConflictsResolutionStrategyEnum resolutionStrategy;
  Function? conflictsResolutionFunction;

  ConflictsHandler({
    this.resolutionStrategy = ConflictsResolutionStrategyEnum.lastWriterWins,
    this.conflictsResolutionFunction,
  }) {
    if (resolutionStrategy == ConflictsResolutionStrategyEnum.predefinedRules &&
        conflictsResolutionFunction == null) {
      throw Exception(
          "conflictsResolutionFunction is required when you put ConflictsResolutionStrategyEnum.PredefinedRules");
    }
  }

  @override
  ConflictsResolutionStrategyEnum getConflictsResolutionStrategy() {
    return resolutionStrategy;
  }

  @override
  Future<ISyncableObject> resolveConflicts(
      ISyncableObject oblObject, ISyncableObject newObject) async {
    if (resolutionStrategy == ConflictsResolutionStrategyEnum.lastWriterWins) {
      return newObject;
    }

    if (resolutionStrategy ==
        ConflictsResolutionStrategyEnum.timestampOrdering) {
      return oblObject.metadata.timestamp > newObject.metadata.timestamp
          ? oblObject
          : newObject;
    }

    if (resolutionStrategy == ConflictsResolutionStrategyEnum.predefinedRules) {
      if (conflictsResolutionFunction == null) {
        throw Exception(
            "conflictsResolutionFunction is required when you put ConflictsResolutionStrategyEnum.PredefinedRules");
      }

      final resolvedObject =
          await conflictsResolutionFunction?.call(oblObject, newObject);
      resolvedObject.metadata = newObject.metadata;
      return resolvedObject;
    }

    throw Exception("Unsupported resolution strategy");
  }
}
